# Fruitpromotie
Een supermarkt wil mensen naar de winkel lokken met een sterke promotie voor fruit.

Omdat ze op die promotie geen winst maken willen
ze dat compenseren door vlak ernaast een <mark>ander soort fruit te
leggen die ze iets doen stijgen in prijs,</mark>
zodat de winstmarge daarop het verlies gedeeltelijk kan compenseren.

De winkel wil weten <mark>welk fruit ze in promotie moeten zetten</mark> en
welk fruit het <mark>meeste kans maakt om bij het fruit in promotie gekocht te worden.</mark>

---

1. Maak gebruik van de top 3 fruitvoorkeuren (fruitaankoop.csv) van de studenten TI om de regels op te stellen.
Deze dataset bevat voor elke student zijn drie favoriete fruiten (op zichzelf in willekeurige volgorde).

In [60]:
import pandas as pd


transacties = pd.read_csv('fruitaankoop.csv',sep=';',index_col=0)
transacties



Aardbei  Ananas  Appel  Banaan  Kers  Kiwi  Meloen  Peer  Pruim  \
 Student001        0       1      0       0     1     1       0     0      0   
 Student002        1       1      1       0     0     0       0     0      0   
 Student003        1       1      0       0     1     0       0     0      0   
 Student004        0       1      0       1     0     0       0     0      0   
 Student005        0       0      1       1     0     0       0     1      0   
...              ...     ...    ...     ...   ...   ...     ...   ...    ...   
 Student246        1       0      0       1     0     0       1     0      0   
 Student247        1       0      1       1     0     0       0     0      0   
 Student248        0       0      1       1     0     0       0     1      0   
 Student249        0       0      1       1     0     0       0     1      0   
 Student250        1       0      0       1     0     0       1     0      0   

             Sinaasappel  
 Student001            0  
 Student002            0  
 Student003            0  
 Student004            1  
 Student005            0  
...                  ...  
 Student246            0  
 Student247            0  
 Student248            0  
 Student249            0  
 Student250            0  

[236 rows x 10 columns]

---
2. Maak met deze lijst associatieregels.
Gebruik voor het apriori of fp-growth algoritme volgende parameters:
    * support = 0.1
    * confidence = 0.3
    * minlen=2, maxlen=2


In [89]:
from mlxtend.frequent_patterns import fpgrowth
from mlxtend.frequent_patterns import association_rules

itemsets = fpgrowth(transacties, min_support=0.1, use_colnames=True)
rules = association_rules(itemsets, metric='confidence', min_threshold=0.3)

def rule_filter(row, minlen, maxlen):
    length = len(row['antecedents']) + len(row['consequents'])
    return minlen <= length <= maxlen

result = rules[rules.apply(lambda row: rule_filter(row, 2, 2), axis=1)].sort_values(by='confidence', ascending=False)
result

## Wanneer iemand een Ananas heeft gekocht dan is de kans dat
# die een Aardbei in zijn kar legt 74% -> de confidence is 0.740 de lift is 1.02
#

antecedents consequents  antecedent support  consequent support   support  \
1     (Ananas)   (Aardbei)            0.343220            0.724576  0.254237   
5       (Kiwi)   (Aardbei)            0.165254            0.724576  0.118644   
7      (Appel)   (Aardbei)            0.461864            0.724576  0.313559   
4       (Kers)   (Aardbei)            0.237288            0.724576  0.152542   
14    (Banaan)   (Aardbei)            0.508475            0.724576  0.322034   
16    (Meloen)   (Aardbei)            0.233051            0.724576  0.131356   
15    (Meloen)    (Banaan)            0.233051            0.508475  0.105932   
13   (Aardbei)    (Banaan)            0.724576            0.508475  0.322034   
9      (Appel)    (Banaan)            0.461864            0.508475  0.203390   
6    (Aardbei)     (Appel)            0.724576            0.461864  0.313559   
8     (Banaan)     (Appel)            0.508475            0.461864  0.203390   
2     (Ananas)     (Appel)            0.343220            0.461864  0.131356   
0    (Aardbei)    (Ananas)            0.724576            0.343220  0.254237   
3     (Ananas)    (Banaan)            0.343220            0.508475  0.105932   

    confidence      lift  leverage  conviction  
1     0.740741  1.022309  0.005548    1.062349  
5     0.717949  0.990853 -0.001095    0.976502  
7     0.678899  0.936960 -0.021097    0.857748  
4     0.642857  0.887218 -0.019391    0.771186  
14    0.633333  0.874074 -0.046395    0.751156  
16    0.563636  0.777884 -0.037507    0.631179  
15    0.454545  0.893939 -0.012568    0.901130  
13    0.444444  0.874074 -0.046395    0.884746  
9     0.440367  0.866055 -0.031456    0.878300  
6     0.432749  0.936960 -0.021097    0.948672  
8     0.400000  0.866055 -0.031456    0.896893  
2     0.382716  0.828633 -0.027165    0.871780  
0     0.350877  1.022309  0.005548    1.011796  
3     0.308642  0.606996 -0.068587    0.710956

---
3. Zoek de associatieregel met de hoogste confidence waarvan de lift ook voldoende is.
<br>
<br>
<b>Annanas -> Aardbei
<br>
Wanneer iemand een ananas heeft gekocht dan is de kans dat die ook een Aardbei in zijn
 kar legt 74%</b>

---
4. Welk fruit zal de winkel op basis van die regel in promotie zetten?
<br>
<br>
<b>Annanas en de prijs van de aarbei laten stijgen</b>

---
5. Welk fruit zal de winkel op basis van die regel naast het promotieartikel plaatsen?
<br>
<b>Aarbei</b>

---
6. Hoeveel procent van de mensen die de vragenlijst hebben ingevuld hebben de combinatie
 van de twee fruittypes in hun top 3 staan?
<br>
<b>Hoeveel keer komen annanas en aarbeid samen voor?</b>
<b>25% (0.254237)</b>

In [94]:
from mlxtend.frequent_patterns import fpgrowth
from mlxtend.frequent_patterns import association_rules

itemsets = fpgrowth(transacties, min_support=0.1, use_colnames=True)
itemsets

support                  itemsets
0   0.343220                  (Ananas)
1   0.237288                    (Kers)
2   0.165254                    (Kiwi)
3   0.724576                 (Aardbei)
4   0.461864                   (Appel)
5   0.508475                  (Banaan)
6   0.161017             (Sinaasappel)
7   0.135593                    (Peer)
8   0.233051                  (Meloen)
9   0.254237         (Aardbei, Ananas)
10  0.131356           (Ananas, Appel)
11  0.105932          (Ananas, Banaan)
12  0.152542           (Aardbei, Kers)
13  0.118644           (Aardbei, Kiwi)
14  0.313559          (Aardbei, Appel)
15  0.203390           (Banaan, Appel)
16  0.105932  (Aardbei, Banaan, Appel)
17  0.322034         (Aardbei, Banaan)
18  0.105932          (Meloen, Banaan)
19  0.131356         (Aardbei, Meloen)

---
7. Wat kan je op basis van de lift zeggen over het fruit in promotie?
<br>
Dat dit een zeer goede regel is, want de lift bevestig namelijk onze regel waneer hij groter is dan 1